sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -318,17 +318,30 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
 
   /**
    * Statically evaluate an expression containing zero or more placeholders, given a set
-   * of bindings for placeholder values.
+   * of bindings for placeholder values, if the expression is evaluable. If it is not,
+   * bind statically evaluated expression results to an expression.
    */
-  private def evalExpr(expr: Expression, bindings: Map[ExprId, Option[Any]]) : Option[Any] = {
+  private def bindingExpr(
+      expr: Expression,
+      bindings: Map[ExprId, Option[Expression]]): Option[Expression] = {
     val rewrittenExpr = expr transform {
       case r: AttributeReference =>
         bindings(r.exprId) match {
-          case Some(v) => Literal.create(v, r.dataType)
+          case Some(v) => v
           case None => Literal.default(NullType)
         }
     }
-    Option(rewrittenExpr.eval())
+    if (rewrittenExpr.find(_.isInstanceOf[PythonUDF]).isDefined) {
+      // SPARK-28441: `PythonUDF` can't be statically evaluated.
```

Comment:
many expressions can't be statically evaluated, why only special-case python udf?

Reply:
This issue was found by PythonUDF. I think of covering all unevaluable expressions here, but not sure if it is too aggressive.

Reply:
Do you think we should cover all unevaluable?

Reply:
How do you define "can't be statically evaluated"? Do you mean `!expr.foldable`?

Reply:
PythonUDF is `Unevaluable`. So you can't call `eval` on it.

Reply:
we can't call `eval` if the expression is not foldable, right?

Reply:
here it fakes the empty input case, and evaluate the expressions in subquery. So it doesn't require foldable.

Reply:
We can't call `Expression.eval(null)` if it's not foldable, otherwise exception may be thrown:
1. `AttributeReference.eval(null)` fails with NPE
2. `Nondeterministic.eval(null)` fails because it needs to be initialized first

Whatever hack we use, I'd expect it makes the expression foldable.

Reply:
For 1, AttributeReference was replaced with pre-evaluated value, if it comes from aggregate function. It uses default value. It fakes empty input case.

For 2, I think it is potential issue.

Yeah, here the hack looks like foldable expression. It simulates empty input.

Reply:
It seems necessary to me to check foldable before calling `.eval()`, otherwise there is no guarantee that `.eval()` can success.

Reply:
Good point. Use `foldable` here to check.